### PR TITLE
feat: create JSON schema for v1alpha1.Config

### PR DESCRIPTION
**Pull request by utkuozdemir**
Extend `docgen` tool to generate a JSON schema for `v1alpha1.Config` if a new optional cli arg is provided.

Extend the YAML-structured code comments on config fields to allow overriding the generated schema.

Add custom schemas for complex types.

Related to siderolabs/talos#6705.

Signed-off-by: Utku Ozdemir <utku.ozdemir@siderolabs.com>

### CODE REVIEW

File: pkg/machinery/config/types/v1alpha1/v1alpha1_types.go [hunk 1]
```diff
@@ -1454,6 +1504,9 @@ type TimeConfig struct {
 	//     Specifies the timeout when the node time is considered to be in sync unlocking the boot sequence.
 	//     NTP sync will be still running in the background.
 	//     Defaults to "infinity" (waiting forever for time sync)
+	//   schema:
+	//     type: string
+	//     pattern: ^[-+]?(((\d+(\.\d*)?|\d*(\.\d+)+)(ns|us|µs|ms|s|m|h))|0)+$
```
Comment (utkuozdemir) [inline pkg/machinery/config/types/v1alpha1/v1alpha1_types.go:1509]: Test of this pattern: https://regex101.com/r/xNTblv/2

Comment (frezbo) [reply pkg/machinery/config/types/v1alpha1/v1alpha1_types.go:1509]: that's a lot of matches :cool:

Comment (utkuozdemir) [reply pkg/machinery/config/types/v1alpha1/v1alpha1_types.go:1509]: yes, took them from Go's internal parse tests. It matches all except the ones fail due to overflow.

File: pkg/machinery/config/types/v1alpha1/v1alpha1_config_schema_test.go [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+var (
+	//go:embed v1alpha1_config.schema.json
+	schemaData string
+
+	//go:embed testdata/schema/v1alpha1_config.tmpl.yaml
+	configTemplateData []byte
+)
+
+func TestSchemaValidation(t *testing.T) {
+	schema, err := validatejsonschema.CompileString("https://www.talos.dev/schemas/v1alpha1/config.json", schemaData)
```
Comment (smira) [inline pkg/machinery/config/types/v1alpha1/v1alpha1_config_schema_test.go:36]: my only concern with this test is how we can keep it up to date, as it is based on static machine config. 

can we probably try to use `generate` package here and generate the machine config on the fly with the test, patching it as necessary for each test case?

**Review by smira — APPROVED** [review]
hard to validate the result lol, but this looks great to me! :cool:

File: hack/docgen/schema.go [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+		Ref:     "#/$defs/Config",
+	}
+
+	definitions := make(map[string]*jsonschema.Schema, len(doc.Structs))
+
+	for _, st := range doc.Structs {
+		definitions[st.Name] = structToSchema(st)
+	}
+
+	schema.Definitions = definitions
```
Comment (DmitriyMV) [inline hack/docgen/schema.go:207]: I think we can use [slices.ToMap](https://github.com/siderolabs/gen/blob/main/slices/slices.go#L89) here directly

File: hack/docgen/schema.go [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+type SchemaTypeInfo struct {
+	typeName string
+	ref      string
+}
+
+type SchemaDefinitionInfo struct {
+	typeInfo           SchemaTypeInfo
+	arrayItemsTypeInfo *SchemaTypeInfo
+	mapValueTypeInfo   *SchemaTypeInfo
+	enumValues         []any
```
Comment (DmitriyMV) [inline hack/docgen/schema.go:54]: Any reason for `[]any` and not `[]string`?

Comment (DmitriyMV) [reply hack/docgen/schema.go:54]: Ah - assigning to  to `jsonschema.Schema.Enum`.

File: hack/docgen/schema.go [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+
+	if typeInfo.typeName != "" {
+		schema.Type = typeInfo.typeName
+	}
+
+	if typeInfo.ref != "" {
+		schema.Ref = typeInfo.ref
+	}
+
+	return schema
```
Comment (DmitriyMV) [inline hack/docgen/schema.go:116]: One style here and below will look better I think  - either `&schema` everywhere or `&jsonschema.Schema{}`.

**Review by DmitriyMV — APPROVED** [review]

File: hack/docgen/schema.go (hunk 3 above)
Comment (utkuozdemir) [reply hack/docgen/schema.go]: done

File: hack/docgen/schema.go (hunk 5 above)
Comment (utkuozdemir) [reply hack/docgen/schema.go]: good point, done

**Pull request merged**